tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.cc

```diff
@@ -509,6 +509,7 @@ Status TrtShapeOptimizationProfile::RestoreProfiles(
   int K = n_bindings / n_profiles;
 #endif
   int n_inputs = GetNumberOfEngineInputs(engine);
+  need_profiles_ &= n_inputs > 0;
```

Comment:
Can you explain in what situation getNBOptimizationProfiles() is 0 but n_inputs > 0?
On the other hand, adding this line will make the path where RestoreProfiles() is executed different from the path when we build profile and don't execute RestoreProfiles(). I meant, if there is no profiles, but n_inputs > 0, then one path has need_profiles_ true but another path has need_profiles_ false.

Reply:
> Can you explain in what situation getNBOptimizationProfiles() is 0 but n_inputs > 0?

The question is other way around, is there a case where getNBOptimizationProfiles() > 0 but n_inputs==0?
Yes. Assume that we have a single input tensor x to INetworkDefinition, and we define one profile with min=opt=max. Assume further that only the shape of x is used. TRT 7 in this case would prune the input, so the engine itself would have 0 inputs. 

> On the other hand, adding this line will make the path where RestoreProfiles() is executed different from the path when we build profile and don't execute RestoreProfiles(). 

Valid point (but again other way around: we are concerned with the case when n_inputs == 0 && n_profiles > 0). 

I will fix this.

Reply:
- When we convert the engine, then profiles min, opt, max vectors are initialized from the input shapes. We will have `min.size() == 2*ctx->num_inputs()` [ref](https://github.com/tensorflow/tensorflow/blob/f0014080fb29a9e4038254019f105f0108565e5d/tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.h#L79).
- When we restore the engines then `min.size() == 2 * n_inputs`, where `n_inputs` is the number of input bindings. This can differ from `ctx->num_inputs()` if an input was pruned. This can happen only in TRT 7. TRT 8 does not prune inputs.
- This shows that the profile vectors (min, max, opt) differ on the two code paths.
- The line `need_profiles_ &= n_inputs > 0;` introduces another difference in the code path, with the aim to fix profile selection (disable it), because profile selection would [fail here](https://github.com/tensorflow/tensorflow/blob/f0014080fb29a9e4038254019f105f0108565e5d/tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.h#L117) if an input was pruned and the engine was restored.
- This solves the case when there are no inputs left for the engine.
- There can be a case where `n_inputs > 0`, but `n_inputs != ctx->num_inputs()` , I expect that such a case would still fail. A proper implementation would probably need to modify `IncludesShapes` so that we map `ctx->input` to the `binding_idx`, or modify the `RestoreProfiles` to set some placeholder values for the pruned inputs.
- I will prepare a separate PR with that fix.
- The unit test added in this PR does not trigger the pruned input case. Therefore I can remove the `need_profiles_ &= n_inputs > 0;` condition and add a proper fix in a follow up PR.

Reply:
Removed this line.